File: arrays/romanToInt.go
package main

/*

link problem leetcode -> https://leetcode.com/problems/two-sum/
time complexity  -> O(n)
space complexity -> O(1)

*/
func romanToInt(s string) int {
	subtraction := buildMap()
	values := buildValues()

	if len(s) == 1{
		return values[s]
	}else if len(s) == 2{
		if val, ok := subtraction[string(s[0]) + string(s[1])]; ok {
			return val
		}
		return values[string(s[0])] + values[string(s[1])]
	}

	sum := 0

	actual := len(s) - 1
	next := len(s) - 2 

	for next >= 0 && actual >= 0 {
		if val, ok := subtraction[string(s[next]) + string(s[actual])]; ok {
			sum += val
			next -= 2
			actual -=2 
		}else{
			sum += values[string(s[actual])] 
			next -= 1
			actual -= 1 
		}
	}


	if actual == 0 {
		sum += values[string(s[actual])]
	}

	
	return sum

}

func buildValues() map[string]int{
	mymap := make(map[string]int)
	mymap["I"] = 1
	mymap["V"] = 5
	mymap["X"] = 10
	mymap["L"] = 50
	mymap["C"] = 100
	mymap["D"] = 500
	mymap["M"] = 1000

	return mymap
}

func buildMap() map[string]int {
	mymap := make(map[string]int)
	mymap["IV"] = 4
	mymap["IX"] = 9
	mymap["XL"] = 40
	mymap["XC"] = 90
	mymap["CD"] = 400
	mymap["CM"] = 900

	return mymap
}

File: arrays_and_strings/romanToInt.py
#Link to problem: https://leetcode.com/problems/roman-to-integer/description/
#Time complexity: O(n)
#Space complexity: O(1)

class Solution:
    def romanToInt(self, s: str) -> int:
        romanToInteger = {"I":1,"V":5,"X":10,"L":50,"C":100,"D":500,"M":1000,"IV":4,"IX":9,"XL":40,"XC":90,"CD":400,"CM":900}

        actual = ""
        sum = 0
        i = 0
        while i < len(s) - 1:
            actual = s[i]+s[i+1]
            combination = romanToInteger.get(actual)
            if combination is not None:
                sum += combination
                i += 2
            else:
                sum += romanToInteger.get(s[i])
                i += 1
        
        lastPair = romanToInteger.get(s[len(s) - 2] + s[len(s) - 1])
        if lastPair is None:
            sum += romanToInteger.get(s[len(s) - 1])

        return sum



solution = Solution()
print(solution.romanToInt("III"))
print(solution.romanToInt("LVIII"))
print(solution.romanToInt("MCMXCIV"))